base/compiler/EscapeAnalysis/EscapeAnalysis.jl

```diff
@@ -0,0 +1,1389 @@
+  is slightly lower than `NoEscape`, but at the same time doesn't represent any meaning
+  other than it's not analyzed yet (thus it's not formally part of the lattice).
+"""
+struct EscapeLattice
+    Analyzed::Bool
+    ReturnEscape::Bool
+    ThrownEscape::Bool
+    EscapeSites::BitSet
+    AliasEscapes #::Union{FieldEscapes,ArrayEscapes,Bool}
+    # TODO: ArgEscape::Int
+
+    function EscapeLattice(
+        Analyzed::Bool,
+        ReturnEscape::Bool,
+        ThrownEscape::Bool,
+        EscapeSites::BitSet,
+        AliasEscapes#=::Union{FieldEscapes,ArrayEscapes,Bool}=#,
+        )
+        @nospecialize AliasEscapes
+        return new(
+            Analyzed,
+            ReturnEscape,
+            ThrownEscape,
+            EscapeSites,
+            AliasEscapes,
+            )
+    end
+    function EscapeLattice(
+        x::EscapeLattice,
+        # non-concrete fields should be passed as default arguments
+        # in order to avoid allocating non-concrete `NamedTuple`s
+        AliasEscapes#=::Union{FieldEscapes,ArrayEscapes,Bool}=# = x.AliasEscapes;
+        Analyzed::Bool = x.Analyzed,
+        ReturnEscape::Bool = x.ReturnEscape,
+        ThrownEscape::Bool = x.ThrownEscape,
+        EscapeSites::BitSet = x.EscapeSites,
+        )
+        @nospecialize AliasEscapes
+        return new(
+            Analyzed,
+            ReturnEscape,
+            ThrownEscape,
+            EscapeSites,
+            AliasEscapes,
+            )
+    end
+end
+
+# precomputed default values in order to eliminate computations at each callsite
+const BOT_ESCAPE_SITES = BitSet()
+const ARGUMENT_ESCAPE_SITES = BitSet(0)
+const TOP_ESCAPE_SITES = BitSet(0:100_000)
+
+const BOT_ALIAS_ESCAPES = false
+const TOP_ALIAS_ESCAPES = true
+
+# the constructors
+NotAnalyzed() = EscapeLattice(false, false, false, BOT_ESCAPE_SITES, BOT_ALIAS_ESCAPES) # not formally part of the lattice
+NoEscape() = EscapeLattice(true, false, false, BOT_ESCAPE_SITES, BOT_ALIAS_ESCAPES)
+ReturnEscape(pc::Int) = EscapeLattice(true, true, false, BitSet(pc), BOT_ALIAS_ESCAPES)
+ThrownEscape(pc::Int) = EscapeLattice(true, false, true, BitSet(pc), BOT_ALIAS_ESCAPES)
+ArgumentReturnEscape() = EscapeLattice(true, true, false, ARGUMENT_ESCAPE_SITES, TOP_ALIAS_ESCAPES) # TODO allow interprocedural field analysis?
+AllEscape() = EscapeLattice(true, true, true, TOP_ESCAPE_SITES, TOP_ALIAS_ESCAPES)
+
+# Convenience names for some ⊑ queries
+has_not_analyzed(x::EscapeLattice) = x == NotAnalyzed()
+has_no_escape(x::EscapeLattice) = ignore_aliasescapes(x) ⊑ NoEscape()
+has_return_escape(x::EscapeLattice) = x.ReturnEscape
+has_return_escape(x::EscapeLattice, pc::Int) = has_return_escape(x) && pc in x.EscapeSites
+has_thrown_escape(x::EscapeLattice) = x.ThrownEscape
+has_thrown_escape(x::EscapeLattice, pc::Int) = has_thrown_escape(x) && pc in x.EscapeSites
+has_only_throw_escape(x::EscapeLattice, pc::Int) = has_thrown_escape(x, pc) && length(x.EscapeSites) == 1
```

Comment:
For now let's make a PR to the EscapeAnalysis.jl repository in order to make it easier to track diffs (I agree that it's not very fun to work on both code, but I'm still trying to find a way to nicely share the code in EA.jl and the code here).

Reply:
Resolved due to going a different direction with handling BoundsError escapes